In [6]:
from __future__ import print_function
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K
from keras.layers import Activation
from keras.layers import regularizers
from keras.layers import BatchNormalization
from prettytable import PrettyTable

In [2]:
# Credits: https://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py

batch_size = 128
num_classes = 10
epochs = 12

# input image dimensions
img_rows, img_cols = 28, 28

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples


In [3]:
# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

In [4]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

W0811 17:59:04.471589  7896 deprecation_wrapper.py:119] From C:\Users\hp\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:74: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.

W0811 17:59:04.487215  7896 deprecation_wrapper.py:119] From C:\Users\hp\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

W0811 17:59:04.487215  7896 deprecation_wrapper.py:119] From C:\Users\hp\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.

W0811 17:59:04.643455  7896 deprecation_wrapper.py:119] From C:\Users\hp\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:3976: The name tf.nn.max_pool is deprecated. Please use tf.nn.max_pool2d instead.

W0811 17:59:04.643455  7896 deprecation_wrapper.py:119] From C:\Users\hp\Anaconda3\lib\site

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 113s 2ms/step - loss: 0.2674 - acc: 0.9171 - val_loss: 0.0611 - val_acc: 0.9786
Epoch 2/12
60000/60000 [==============================] - 111s 2ms/step - loss: 0.0871 - acc: 0.9742 - val_loss: 0.0517 - val_acc: 0.9828
Epoch 3/12
60000/60000 [==============================] - 111s 2ms/step - loss: 0.0658 - acc: 0.9798 - val_loss: 0.0360 - val_acc: 0.9883
Epoch 4/12
60000/60000 [==============================] - 109s 2ms/step - loss: 0.0521 - acc: 0.9847 - val_loss: 0.0349 - val_acc: 0.9887
Epoch 5/12
60000/60000 [==============================] - 111s 2ms/step - loss: 0.0454 - acc: 0.9863 - val_loss: 0.0309 - val_acc: 0.9896
Epoch 6/12
60000/60000 [==============================] - 111s 2ms/step - loss: 0.0418 - acc: 0.9874 - val_loss: 0.0261 - val_acc: 0.9917
Epoch 7/12
60000/60000 [==============================] - 112s 2ms/step - loss: 0.0366 - acc: 0.9887 - val_loss: 0.0286 - 

# Assignment

We have to make about 3 distinct architectures of convolutional netwroks.

### Architecture 1:

In [7]:
model_1 = Sequential()
model_1.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model_1.add(Conv2D(32, (3, 3), activation='relu'))
model_1.add(MaxPooling2D(pool_size=(3, 3)))
model_1.add(BatchNormalization())
model_1.add(Dropout(0.5))
model_1.add(Flatten())
model_1.add(Dense(64, activation='relu'))
model_1.add(Dense(num_classes, activation='softmax'))

model_1.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

model_1.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score_1 = model_1.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score_1[0])
print('Test accuracy:', score_1[1])

W0811 18:35:14.560059  7896 deprecation_wrapper.py:119] From C:\Users\hp\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:1834: The name tf.nn.fused_batch_norm is deprecated. Please use tf.compat.v1.nn.fused_batch_norm instead.



Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 73s 1ms/step - loss: 0.1829 - acc: 0.9419 - val_loss: 0.0568 - val_acc: 0.9823
Epoch 2/12
60000/60000 [==============================] - 80s 1ms/step - loss: 0.0683 - acc: 0.9788 - val_loss: 0.0425 - val_acc: 0.9857
Epoch 3/12
60000/60000 [==============================] - 89s 1ms/step - loss: 0.0523 - acc: 0.9841 - val_loss: 0.0399 - val_acc: 0.9878
Epoch 4/12
60000/60000 [==============================] - 78s 1ms/step - loss: 0.0447 - acc: 0.9857 - val_loss: 0.0315 - val_acc: 0.9904
Epoch 5/12
60000/60000 [==============================] - 79s 1ms/step - loss: 0.0367 - acc: 0.9883 - val_loss: 0.0325 - val_acc: 0.9895
Epoch 6/12
60000/60000 [==============================] - 77s 1ms/step - loss: 0.0325 - acc: 0.9894 - val_loss: 0.0351 - val_acc: 0.9887
Epoch 7/12
60000/60000 [==============================] - 77s 1ms/step - loss: 0.0302 - acc: 0.9902 - val_loss: 0.0338 - val_acc

### Arhcitecture 2

In [9]:
model_2 = Sequential()
model_2.add(Conv2D(32, kernel_size=(5, 5),
                 activation='relu',
                 input_shape=input_shape))
model_2.add(Conv2D(64, (3, 3), activation='relu'))
model_2.add(BatchNormalization())
model_1.add(Dropout(0.1))
model_2.add(Conv2D(64, (2, 2), activation='relu'))
model_2.add(MaxPooling2D(pool_size=(3, 3)))
model_2.add(BatchNormalization())
model_2.add(Flatten())
model_2.add(Dense(64, activation='relu'))
model_2.add(BatchNormalization())
model_2.add(Dropout(0.25))
model_2.add(Dense(num_classes, activation='softmax'))

model_2.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adam(),
              metrics=['accuracy'])

model_2.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score_2 = model_2.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score_2[0])
print('Test accuracy:', score_2[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 303s 5ms/step - loss: 0.1254 - acc: 0.9651 - val_loss: 0.0680 - val_acc: 0.9776
Epoch 2/12
60000/60000 [==============================] - 303s 5ms/step - loss: 0.0431 - acc: 0.9879 - val_loss: 0.0437 - val_acc: 0.9840
Epoch 3/12
60000/60000 [==============================] - 302s 5ms/step - loss: 0.0320 - acc: 0.9904 - val_loss: 0.0273 - val_acc: 0.9919
Epoch 4/12
60000/60000 [==============================] - 300s 5ms/step - loss: 0.0230 - acc: 0.9928 - val_loss: 0.0477 - val_acc: 0.9852
Epoch 5/12
60000/60000 [==============================] - 305s 5ms/step - loss: 0.0198 - acc: 0.9941 - val_loss: 0.0386 - val_acc: 0.9867
Epoch 6/12
60000/60000 [==============================] - 328s 5ms/step - loss: 0.0168 - acc: 0.9945 - val_loss: 0.0397 - val_acc: 0.9879
Epoch 7/12
60000/60000 [==============================] - 332s 6ms/step - loss: 0.0143 - acc: 0.9953 - val_loss: 0.0364 - 

### Architecture 3:

In [10]:
model_3 = Sequential()
model_3.add(Conv2D(32, kernel_size=(2, 2),
                 activation='sigmoid',
                 input_shape=input_shape))
model_3.add(Conv2D(64, (3, 3), activation='sigmoid'))
model_3.add(BatchNormalization())
model_3.add(MaxPooling2D(pool_size=(3, 3)))
model_3.add(Dropout(0.5))
model_3.add(Flatten())
model_3.add(Dense(128, activation='sigmoid'))
model_3.add(BatchNormalization())
model_3.add(Dropout(0.2))
model_3.add(Dense(num_classes, activation='softmax'))

model_3.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adagrad(),
              metrics=['accuracy'])

model_3.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score_3 = model_3.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score_3[0])
print('Test accuracy:', score_3[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 316s 5ms/step - loss: 0.1750 - acc: 0.9469 - val_loss: 0.4786 - val_acc: 0.8493
Epoch 2/12
60000/60000 [==============================] - 319s 5ms/step - loss: 0.0886 - acc: 0.9730 - val_loss: 2.1779 - val_acc: 0.6131
Epoch 3/12
60000/60000 [==============================] - 317s 5ms/step - loss: 0.0738 - acc: 0.9772 - val_loss: 0.1461 - val_acc: 0.9577
Epoch 4/12
60000/60000 [==============================] - 327s 5ms/step - loss: 0.0647 - acc: 0.9802 - val_loss: 0.5562 - val_acc: 0.8509
Epoch 5/12
60000/60000 [==============================] - 326s 5ms/step - loss: 0.0578 - acc: 0.9822 - val_loss: 0.1194 - val_acc: 0.9633
Epoch 6/12
60000/60000 [==============================] - 324s 5ms/step - loss: 0.0552 - acc: 0.9826 - val_loss: 0.0465 - val_acc: 0.9847
Epoch 7/12
60000/60000 [==============================] - 324s 5ms/step - loss: 0.0515 - acc: 0.9846 - val_loss: 0.2179 - 

### Architecture 4:

In [5]:
# Activation = selu & softmax , without Batch normalization and dropout, Initializer = kernel_initializer &
# bias_initializer, Optimizer = Adagrad
epochs = 10
batch_size = 256

model_4 = Sequential()
model_4.add(Conv2D(32, kernel_size=(5, 5),
                 activation='tanh',
                 input_shape=input_shape))
model_4.add(Dense(45,
                kernel_initializer='random_uniform'))
model_4.add(Conv2D(64, (4, 4), activation='softmax'))
model_4.add(BatchNormalization())
model_4.add(Dropout(0.25))
model_4.add(MaxPooling2D(pool_size=(3, 3)))
model_4.add(Flatten())
model_4.add(Dense(128, activation='sigmoid'))
model_4.add(BatchNormalization())
model_4.add(Dropout(0.5))
model_4.add(Dense(num_classes, activation='softmax'))

model_4.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adamax(),
              metrics=['accuracy'])

model_4.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score_4 = model_4.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score_4[0])
print('Test accuracy:', score_4[1])

W0811 22:37:34.260540  6576 deprecation_wrapper.py:119] From C:\Users\hp\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:174: The name tf.get_default_session is deprecated. Please use tf.compat.v1.get_default_session instead.

W0811 22:37:34.260540  6576 deprecation_wrapper.py:119] From C:\Users\hp\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:181: The name tf.ConfigProto is deprecated. Please use tf.compat.v1.ConfigProto instead.

W0811 22:37:34.338661  6576 deprecation_wrapper.py:119] From C:\Users\hp\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:1834: The name tf.nn.fused_batch_norm is deprecated. Please use tf.compat.v1.nn.fused_batch_norm instead.

W0811 22:37:34.416781  6576 deprecation.py:506] From C:\Users\hp\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:3445: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.
Instructions for updating:
Pl

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 354s 6ms/step - loss: 0.1780 - acc: 0.9479 - val_loss: 0.0673 - val_acc: 0.9795
Epoch 2/10
60000/60000 [==============================] - 378s 6ms/step - loss: 0.0719 - acc: 0.9778 - val_loss: 0.0656 - val_acc: 0.9792
Epoch 3/10
60000/60000 [==============================] - 360s 6ms/step - loss: 0.0588 - acc: 0.9825 - val_loss: 0.0538 - val_acc: 0.9835
Epoch 4/10
60000/60000 [==============================] - 368s 6ms/step - loss: 0.0494 - acc: 0.9842 - val_loss: 0.0468 - val_acc: 0.9853
Epoch 5/10
60000/60000 [==============================] - 366s 6ms/step - loss: 0.0438 - acc: 0.9871 - val_loss: 0.0588 - val_acc: 0.9811
Epoch 6/10
60000/60000 [==============================] - 362s 6ms/step - loss: 0.0389 - acc: 0.9884 - val_loss: 0.0463 - val_acc: 0.9860
Epoch 7/10
60000/60000 [==============================] - 397s 7ms/step - loss: 0.0340 - acc: 0.9893 - val_loss: 0.0412 - 

## Conclusions

In [11]:
number= [1,2,3,4] 
model = ["Architecture 1","Architecture 2","Architecture 3","Architecture 4"] 
opt   = ['Adadelta', 'Adam', 'Adagrad', 'Adamax']
act   = ['relu', 'relu', 'sigmoid', 'tanh,sigmoid']
loss  = [0.028312752436022673,0.03223905730191964,0.03459447893754113,score_4[0]] 
acc   = [0.9919,0.9913,0.9881,score_4[1]]       

#Initialize Prettytable 
pt = PrettyTable() 
pt.add_column("Sr.No.", number) 
pt.add_column("Model", model) 
pt.add_column("Optimizer", opt)
pt.add_column('Activation ', act)
pt.add_column("Test Loss", loss) 
pt.add_column("Test Accuracy", acc) 
print(pt)

+--------+----------------+-----------+--------------+----------------------+---------------+
| Sr.No. |     Model      | Optimizer | Activation   |      Test Loss       | Test Accuracy |
+--------+----------------+-----------+--------------+----------------------+---------------+
|   1    | Architecture 1 |  Adadelta |     relu     | 0.028312752436022673 |     0.9919    |
|   2    | Architecture 2 |    Adam   |     relu     | 0.03223905730191964  |     0.9913    |
|   3    | Architecture 3 |  Adagrad  |   sigmoid    | 0.03459447893754113  |     0.9881    |
|   4    | Architecture 4 |   Adamax  | tanh,sigmoid | 0.036352840721810935 |     0.9884    |
+--------+----------------+-----------+--------------+----------------------+---------------+


Architecture 1 is getting highest accuracy of 99.19% with minimum loss, in which I have used Adadelta optimizer and relu activation.


### I have tried 4 CNN architectures:
 

### 1. with differnt number of layers
### 2. different kernel values: (2,2), (3,3), (5,5)
### 3. different activations: relu, tanh, sigmoid
### 4. different optimizers: adam, adadelta, adagrad, adamax
### 5. adding batchnormalisation layers
### 6. adding some dropout layers, maxpool layers, flattening layers and dense layers